File: 3610-find-x-sum-of-all-k-long-subarrays-i/find-x-sum-of-all-k-long-subarrays-i.java
class Solution {
    public int[] findXSum(int[] nums, int k, int x) {
        int n = nums.length;

        HashMap<Integer,Integer> mp = new HashMap<>();
        for (int i = 0; i < k; i++) {
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
        }

        int[] ans = new int[n - k + 1];

        BiFunction<HashMap<Integer,Integer>, Integer, Integer> sum = (map, X) -> {
            List<Map.Entry<Integer, Integer>> list = new ArrayList<>(map.entrySet());

            Collections.sort(list, (a, b) -> {
                if (a.getValue().equals(b.getValue()))
                    return b.getKey() - a.getKey();
                return b.getValue() - a.getValue();
            });

            int s = 0;
            int count = 0;

            for (Map.Entry<Integer, Integer> e : list) {
                if (count == X) break;
                s += e.getKey() * e.getValue();
                count++;
            }

            return s;
        };

        ans[0] = sum.apply(mp, x);

        for (int i = k; i < n; i++) {
            mp.put(nums[i - k], mp.get(nums[i - k]) - 1);
            
            if (mp.get(nums[i - k]) == 0) mp.remove(nums[i - k]);

            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);

            ans[i - k + 1] = sum.apply(mp, x);
        }

        return ans;
    }
}

File: 3610-find-x-sum-of-all-k-long-subarrays-i/find-x-sum-of-all-k-long-subarrays-i.cpp
class Solution {
public:
    vector<int> findXSum(vector<int>& nums, int k, int x) {
        int n = nums.size();
        vector<int> answer;
        
        for (int i = 0; i <= n - k; i++) {
            answer.push_back(calculateXSum(nums, i, i + k - 1, x));
        }
        
        return answer;
    }
    
private:
    int calculateXSum(vector<int>& nums, int start, int end, int x) {
        // Count occurrences
        unordered_map<int, int> freq;
        for (int i = start; i <= end; i++) {
            freq[nums[i]]++;
        }
        
        // If less than x distinct elements, return sum of array
        if (freq.size() < x) {
            int sum = 0;
            for (int i = start; i <= end; i++) {
                sum += nums[i];
            }
            return sum;
        }
        
        // Create vector of pairs (value, frequency)
        vector<pair<int, int>> elements;
        for (auto& p : freq) {
            elements.push_back({p.first, p.second});
        }
        
        // Sort by frequency (descending), then by value (descending)
        sort(elements.begin(), elements.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
            if (a.second != b.second) {
                return a.second > b.second; // Higher frequency first
            }
            return a.first > b.first; // Higher value first if same frequency
        });
        
        // Calculate x-sum (top x elements)
        int xSum = 0;
        for (int i = 0; i < x && i < elements.size(); i++) {
            xSum += elements[i].first * elements[i].second;
        }
        
        return xSum;
    }
};